tests/unit/PointerArrayEmission.cpp

```diff
@@ -0,0 +1,135 @@
+  auto TypedefCPAA = model::ArrayType::make(TypedefCPA.copy(), 37);
+  Tests.emplace_back(TypedefCPAA, "_typedef_0 *const test[37][20]");
+
+  auto VoidPAACP = model::PointerType::makeConst(VoidPAA.copy(), 8);
+  Tests.emplace_back(VoidPAACP, "void *(*const test)[42][15]");
+  auto VoidCPAACP = model::PointerType::makeConst(VoidCPAA.copy(), 8);
+  Tests.emplace_back(VoidCPAACP, "void *const (*const test)[41][16]");
+  auto IntPAAP = model::PointerType::make(IntPAA.copy(), 8);
+  Tests.emplace_back(IntPAAP, "int32_t const *(*test)[40][17]");
+  auto IntCPAAP = model::PointerType::make(IntCPAA.copy(), 8);
+  Tests.emplace_back(IntCPAAP, "int32_t const *const (*test)[39][18]");
+  auto TypedefPAACP = model::PointerType::makeConst(TypedefPAA.copy(), 8);
+  Tests.emplace_back(TypedefPAACP, "_typedef_0 *(*const test)[38][19]");
+  auto TypedefCPAAP = model::PointerType::make(TypedefCPAA.copy(), 8);
+  Tests.emplace_back(TypedefCPAAP, "_typedef_0 *const (*test)[37][20]");
+
+  auto FinalV = model::ArrayType::make(VoidPAACP.copy(), 1);
+  Tests.emplace_back(FinalV, "void *(*const test[1])[42][15]");
+  auto FinalVC = model::ArrayType::make(VoidCPAACP.copy(), 2);
+  Tests.emplace_back(FinalVC, "void *const (*const test[2])[41][16]");
+  auto FinalI = model::ArrayType::make(IntPAAP.copy(), 3);
+  Tests.emplace_back(FinalI, "int32_t const *(*test[3])[40][17]");
+  auto FinalIC = model::ArrayType::make(IntCPAAP.copy(), 4);
+  Tests.emplace_back(FinalIC, "int32_t const *const (*test[4])[39][18]");
+  auto FinalT = model::ArrayType::make(TypedefPAACP.copy(), 5);
+  Tests.emplace_back(FinalT, "_typedef_0 *(*const test[5])[38][19]");
+  auto FinalTC = model::ArrayType::make(TypedefCPAAP.copy(), 6);
+  Tests.emplace_back(FinalTC, "_typedef_0 *const (*test[6])[37][20]");
+
+  auto Extra1 = model::PointerType::makeConst(FinalIC.copy(), 8);
+  Tests.emplace_back(Extra1,
+                     "int32_t const *const (*(*const test)[4])[39][18]");
+  auto Extra2 = model::PointerType::makeConst(Extra1.copy(), 8);
+  Tests.emplace_back(Extra2,
+                     "int32_t const *const "
+                     "(*(*const *const test)[4])[39][18]");
+  auto Extra3 = model::PointerType::make(Extra2.copy(), 8);
+  Tests.emplace_back(Extra3,
+                     "int32_t const *const "
+                     "(*(*const *const *test)[4])[39][18]");
+  auto Extra4 = model::PointerType::make(Extra3.copy(), 8);
+  Tests.emplace_back(Extra4,
+                     "int32_t const *const "
+                     "(*(*const *const **test)[4])[39][18]");
+  auto Extra5 = model::PointerType::make(Extra4.copy(), 8);
+  Tests.emplace_back(Extra5,
+                     "int32_t const *const "
+                     "(*(*const *const ***test)[4])[39][18]");
+  auto Extra6 = model::PointerType::makeConst(Extra5.copy(), 8);
+  Tests.emplace_back(Extra6,
+                     "int32_t const *const "
+                     "(*(*const *const ****const test)[4])[39][18]");
+
```

Comment:
Done.
```
// Tests whether the way we emit pointers and arrays is reasonable.
// Each layer of the types builds on the previous ones, so the test get more
// complex gradually.
//
// \note the easiest way to understand these is to see it fail: just edit
//       one of the strings and watch what it prints: it'll show what
//       we emitted, what we should have, and how the type in question looks
//       like in the model.
```

Reply:
Something like this?
```
// \note also, notice that there is no special logic surrounding function
//       pointers, it's because we always emit those as typedef, so they always
//       have a concrete name to fall onto, making their emission trivial.
```

Reply:
:+1:

Reply:
Done.